hpx/lcos/latch.hpp

```diff
@@ -0,0 +1,143 @@
+        /// Synchronization: None
+        /// Postconditions: counter_ == count.
+        ///
+        explicit latch(std::ptrdiff_t count)
+        {
+        }
+
+        /// Extension: Create a client side representation for the existing
+        /// \a server#latch instance with the given global id \a id.
+        latch(naming::id_type id)
+          : base_type(id)
+        {}
+
+        /// Extension: Create a client side representation for the existing
+        /// \a server#latch instance with the given global id \a id.
+        latch(hpx::future<naming::id_type> && id)
+          : base_type(id.share())
+        {}
+
+        /// Extension: Create a client side representation for the existing
+        /// \a server#latch instance with the given global id \a id.
+        latch(hpx::shared_future<naming::id_type> const& id)
+          : base_type(id)
+        {}
+
+        ///////////////////////////////////////////////////////////////////////
+
+        /// Decrements counter_ by 1 . Blocks at the synchronization point
+        /// until counter_ reaches 0.
+        ///
+        /// Requires: counter_ > 0.
+        ///
+        /// Synchronization: Synchronizes with all calls that block on this
+        /// latch and with all is_ready calls on this latch that return true.
+        ///
+        /// \throws Nothing.
+        ///
+        void count_down_and_wait()
+        {
+            count_down_and_wait_async().get();
+        }
+
+        /// Decrements counter_ by n. Does not block.
+        ///
+        /// Requires: counter_ >= n and n >= 0.
+        ///
+        /// Synchronization: Synchronizes with all calls that block on this
+        /// latch and with all is_ready calls on this latch that return true .
+        ///
+        /// \throws Nothing.
+        ///
+        void count_down(std::ptrdiff_t n)
+        {
+            count_down_async(n).get();
+        }
+
+        /// Returns: counter_ == 0. Does not block.
+        ///
+        /// \throws Nothing.
+        ///
+        bool is_ready() const BOOST_NOEXCEPT
+        {
+            return is_ready_async().get() ? true : false;
+        }
+
+        /// If counter_ is 0, returns immediately. Otherwise, blocks the
+        /// calling thread at the synchronization point until counter_
+        /// reaches 0.
+        ///
+        /// \throws Nothing.
+        ///
+        void wait() const
+        {
+            return wait_async().get();
+        }
+
+        /// \cond NOINTERNAL
+        // extended API
+        hpx::future<void> count_down_and_wait_async()
+        {
+            lcos::server::latch::set_event_action act;
+            return hpx::async(act, get_gid());
+        }
+
+        hpx::future<void> count_down_async(std::ptrdiff_t n)
+        {
+            lcos::server::latch::set_value_action act;
+            return hpx::async(act, get_gid(), std::move(n));
+        }
+
+        hpx::future<std::ptrdiff_t> is_ready_async() const
```

Comment:
Deriving from base_lco_with_value<bool, std::ptrdiff_t> would solve this API asymmetry:
https://gist.github.com/sithhell/34e94c44ca06f39b27ef